fs/sshfs/sshfs.py

```diff
@@ -155,6 +156,28 @@ def listdir(self, path):  # noqa: D102
         with convert_sshfs_errors('listdir', path):
             return self._sftp.listdir(_path)
 
+    def scandir(self, path, namespaces=None, page=None):  # noqa: D102
+        self.check()
+        _path = self.validatepath(path)
+        _namespaces = namespaces or ()
+        _start, _stop = page or (None, None)
+        try:
+            with convert_sshfs_errors('scandir', path, directory=True):
+                _stat_iter = self._sftp.listdir_iter(_path)
+                for _stat in itertools.islice(_stat_iter, _start, _stop):
+                    yield self._make_info(_stat.filename, _stat, _namespaces)
+        except errors.ResourceNotFound:
+            # When given a bad path to listdir, the sftp client raises IOError
+            # with an errno of ENOENT no matter if the path was missing or was
+            # the wrong type (e.g., a file).  For the fs API, we need to figure
+            # out if it was supposed to be ENOTDIR instead of ENOENT.
+            try:
```

Comment:
For this part, I would suggest the following to be more clear about what we are doing:
```python
            if self.isfile(_path):
                six.raise_from(errors.DirectoryExpected(path), None)
            raise
```

This has two advantages:

- It is more clear that we are treating the path being a file as a special case of `ResourceNotFound`
- `six.raise_from` avoids cluttering the backtrace since the user will not see the temporary `ResourceNotFound`